### DISCUSSION TITLE
Stream receiving messages in ReactPHP (client)

### DISCUSSION THREAD

**Discussion by Rishat7c** [reactphp/reactphp#539] [Q&A]
Stream receiving messages in ReactPHP (client)

Hi all. There is a [RASA](https://github.com/rasahq/rasa) server (external server) and my website in PHP, where a custom widget in JS is located. Everything works perfectly.

The [RASA](https://github.com/rasahq/rasa) documentation states that the RASA service supports streaming messaging. In the request to the RASA API that we are making, we added the stream=true parameter, then the responses should not arrive in a batch at once, but one after another as they arrive. For the user, this will mean that the bot responds instantly.

To implement this solution, we connected ReactPHP.
Nothing has changed. Responses come when the request is completed completely. Tell me, what am I doing wrong?
Please share your thoughts or solutions

I am publishing my solution (it is not working, i.e. the answer comes when the request is fully processed)

```
$client = new Browser();
$start = microtime(true);

$client->requestStreaming(
    'POST',
    'https://192.168.0.25/webhooks/rest/webhook?stream=true',
    [
        'Content-Type' => 'application/json'
    ],
    json_encode([
    'sender' => 'test-rg-07',
    'token' => '1239rogn238hHSx',
    'message' => 'Hello, CJ. Let's go to Los Santos',
])
)->then(function (ResponseInterface $response) use ($start) {
    $body = $response->getBody();
    assert($body instanceof Psr\Http\Message\StreamInterface);
    assert($body instanceof React\Stream\ReadableStreamInterface);

    $body->on('data', function ($chunk) use ($start) {
        echo $chunk;
        $diff = sprintf( '%.6f sec.', microtime( true ) - $start);
        echo "Execution time: $diff"; // Execution time: 00.00015
    });

    $body->on('error', function (Exception $e) {
        echo 'Error: ' . $e->getMessage() . PHP_EOL;
    });

    $body->on('close', function () {
        echo '[DONE]' . PHP_EOL;
    });
}, function (Exception $e) {
    echo 'Error: ' . $e->getMessage() . PHP_EOL;
});
```

**Comment by SimonFrings** [comment]
@Rishat7c Thanks for bringing this up :+1:

The code that you shared above seems to be fine, so it currently seems like this isn't a problem caused by ReactPHP. I'm not quite sure what your expected result is and what you're currently getting back, can you go a bit more into detail? Are you getting one result at the end but expect to receive each chunks individually?

This could also be related to how your application is configured. For instance, if you're using an Apache server, it may be the reason why you're receiving a single answer, as Apache utilizes buffering logic. Another possibility is that RASA may not fully support streaming, or perhaps you're not using it correctly - just speculating here.

Interested in your input on this.

**Comment by Rishat7c** [comment]
Hello @SimonFrings 

My client configuration (OpenServer):
- Apache 2.4
- PHP 7.4

When I access an external RASA API server from my website, I receive an array of messages as a result. Each message arrives with a short time interval (more details on the screenshot). I marked the runtime output in the code and it is displayed only when the server returns a complete response. I want to process each message separately, rather than wait for a complete response from the server. Please tell me, is this really possible to implement using PHP and ReactPHP?

![Image](https://github.com/reactphp/reactphp/assets/23475575/3bc154b7-3039-4927-80f1-02e0b21b0c93)

**Reply by SimonFrings** [reply]
> [...] is this really possible to implement using PHP and ReactPHP?

Yes it is, like I said above, the code example you provided looks fine. This seems to be a problem outside of ReactPHP in this case, so I can make an suggestion what you can check on your side.

Receiving each chunk individually and then printing out the result should work according to your code example above. As you've also described (and also taken from your screenshots), you're currently receiving one answer after 6 seconds from whoever is sending you these responses. Have you checked that the sender is also sending these responses in chunks rather than one giant response?

**Reply by Rishat7c** [reply]
> Have you checked that the sender is also sending these responses in chunks rather than one giant response?

@SimonFrings I contacted the developers of the RASA server. They reported that to exchange messages in a stream, just add the GET parameter to the request ?stream=true

They also showed me the code of how it works in Python. Maybe this will help somehow (I'm attaching a screenshot). Any thoughts?

![image](https://github.com/reactphp/reactphp/assets/23475575/b7212e76-2f8c-4a01-9970-6a004a70328b)

**Reply by SimonFrings** [reply]
@Rishat7c Thanks for the additional input, but unfortunately this doesn't help much. If this is actually a problem with the RASA environment, I'm not sure how much I can assist here. Maybe show the RASA developers your code and they'll be able to guide you in the right direction. It has also been a while since my last response (sorry), so maybe this has been resolved in the meantime. If anything comes up on the ReactPHP side of things, I'm more than happy to help you out :+1:
